dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/openapi/ApiFinalise.java

```diff
@@ -473,24 +475,106 @@ private static String getSharedNameForDeclaringType(Api.Parameter p) {
+                                  endpointByMethod
+                                      .computeIfAbsent(
+                                          method, (Function) o -> new ArrayList<Api.Endpoint>())
+                                      .add(e)));
+
+              for (Map.Entry<RequestMethod, List<Api.Endpoint>> methodAndEndpoints :
+                  endpointByMethod.entrySet()) {
+                toPathsAndOperations(methodAndEndpoints, path);
+              }
             });
   }
 
+  private void toPathsAndOperations(
+      Map.Entry<RequestMethod, List<Api.Endpoint>> methodAndEndpoints, String path) {
+    if (methodAndEndpoints.getValue().size() > 1) {
+      Api.Endpoint nonFragmentEndpoint = getPreferredEndpoint(methodAndEndpoints);
+
+      for (int i = 0; i < methodAndEndpoints.getValue().size(); i++) {
+        if (!methodAndEndpoints.getValue().get(i).equals(nonFragmentEndpoint)) {
+          Api.Endpoint endpoint = methodAndEndpoints.getValue().get(i);
+          if (isMergeable(endpoint, nonFragmentEndpoint)) {
+            nonFragmentEndpoint =
+                mergeEndpoints(nonFragmentEndpoint, endpoint, methodAndEndpoints.getKey());
+          } else {
+            AbstractMap.SimpleEntry<String, EnumMap<RequestMethod, Api.Endpoint>>
+                fragmentPathAndOperation = fragmentise(methodAndEndpoints.getKey(), endpoint, path);
+            api.getEndpoints()
+                .put(fragmentPathAndOperation.getKey(), fragmentPathAndOperation.getValue());
+          }
+        }
+      }
+      api.getEndpoints()
+          .computeIfAbsent(path, s -> new EnumMap<>(RequestMethod.class))
+          .put(methodAndEndpoints.getKey(), nonFragmentEndpoint);
+    } else {
+      api.getEndpoints()
+          .computeIfAbsent(path, s -> new EnumMap<>(RequestMethod.class))
+          .put(methodAndEndpoints.getKey(), methodAndEndpoints.getValue().get(0));
+    }
+  }
+
+  private Api.Endpoint getPreferredEndpoint(
+      Map.Entry<RequestMethod, List<Api.Endpoint>> methodAndEndpoints) {
+    for (int i = 0; i < methodAndEndpoints.getValue().size(); i++) {
+      if (consumesOrProducesJson(methodAndEndpoints.getValue().get(i))) {
+        return methodAndEndpoints.getValue().get(i);
+      }
+    }
+
+    return methodAndEndpoints.getValue().get(0);
+  }
+
+  private AbstractMap.SimpleEntry<String, EnumMap<RequestMethod, Api.Endpoint>> fragmentise(
+      RequestMethod requestMethod, Api.Endpoint endpoint, String path) {
+    EnumMap<RequestMethod, Api.Endpoint> requestMethodObjectEnumMap =
+        new EnumMap<>(RequestMethod.class);
+    requestMethodObjectEnumMap.put(requestMethod, endpoint);
+
+    return new AbstractMap.SimpleEntry<>(
+        path + "#" + endpoint.getName(), requestMethodObjectEnumMap);
+  }
+
+  private boolean isMergeable(Api.Endpoint anEndpoint, Api.Endpoint anotherEndpoint) {
+    return anEndpoint.getParameters().equals(anotherEndpoint.getParameters());
```

Comment:
Have you tested that this can be true at all? I mean is `equals` implemented in a way that this has a chance of being true if they have same list of parameters? What about order?